bluej/src/main/java/bluej/compiler/Job.java

```diff
@@ -50,42 +51,72 @@ record Job(CompileInputFile[] sources, Compiler compiler, CompileObserver observ
+    }
+
     /**
      * Compile this job
+     *
+     * <p>The job might have Java and/or Kotlin sources, so we might need to run both compilers.
+     * To handle dependencies correctly, if there are any Kotlin files,
+     * we run the Kotlin compiler first and feed all the source files to it (both Kotlin and Java).
+     * Kotlin compiler will figure out what it needs if some Kotlin files depend on Java sources.</p>
+     *
+     * <p>The results of Kotlin compilation (class files) need to be saved to the output directory
+     * (whether it's a destination directory or a temporary one) so that Java compiler can find them.
+     * This is needed even for CompileType.ERROR_CHECK_ONLY.
+     * That's why we bring a temporary directory handling logic here.
+     * Once we're done with Java compilation, we can delete the temporary directory if needed.</p>
+     *
      */
     public void compile()
     {
         int compilationSequence = nextCompilationSequence.getAndIncrement();
-
+        CompileInputFile[] sourcesArray = sources.toArray(new CompileInputFile[0]);
         try {
             if(observer != null) {
-                observer.startCompile(sources, reason, type, compilationSequence);
+                observer.startCompile(sourcesArray, reason, type, compilationSequence);
             }
 
-            if(destDir != null) {
-                compiler.setDestDir(destDir);
+            File outputDir = type.keepClasses() ? destDir : Files.createTempDirectory("bluej").toFile();
+            List<File> srcFiles = Utility.mapList(sources, CompileInputFile::getCompileInputFile);
+            boolean successful = true;
+            if (kotlinCompiler != null)
+            {
+                configureCompiler(kotlinCompiler, destDir, bpClassLoader);
+                successful &= kotlinCompiler.compile(srcFiles,
+                        observer, internal, userCompileOptions, fileCharset, type, outputDir);
             }
 
-            compiler.setClasspath(bpClassLoader.getClassPathAsFiles());
-
-            compiler.setBootClassPath(null);
-
-            File[] actualSourceFiles = new File[sources.length];
-            for (int i = 0; i < sources.length; i++)
+            List<File> javaSourceFiles =  Utility.filterList(srcFiles,
+                    file -> file.getName().endsWith(".java"));
+            if (!javaSourceFiles.isEmpty())
             {
-                actualSourceFiles[i] = sources[i].getJavaCompileInputFile();
+                configureCompiler(javaCompiler, destDir, bpClassLoader);
+                successful &= javaCompiler.compile(javaSourceFiles,
+                        observer, internal, userCompileOptions, fileCharset, type, outputDir);
             }
 
-            boolean successful = compiler.compile(actualSourceFiles, observer, internal, userCompileOptions, fileCharset, type);
+            if (!type.keepClasses() && outputDir != null)
+            {
+                outputDir.delete();
```

Comment:
It's a good idea to delete the dir (I see now we're not at the moment!), but I think we may need to empty out the temporary class files in it first?

Reply:
In fact, directory was deleted before, but it was done in the `Compiler` implementation in those cases when class files were not needed. I've simply moved this code to the `Job` because now we need to access the same directory with class files in both the Kotlin and Java compilers. But you're right, we do need to empty it out first. I didn't see it, because it's a temporary directory anyway.